File: backend/src/main/java/com/chat/e2e/backend/api/dto/DTOs.java
package com.chat.e2e.backend.api.dto;

import java.time.Instant;
import java.util.*;

public class DTOs {
    public record DeviceSummary(UUID deviceId, String deviceName, String platform, Instant createdAt, Instant lastSeenAt, boolean revoked) {}

    public record EnrollmentStartRequest(String userHandle) {}
    public record EnrollmentStartResponse(UUID deviceId, String nonce, Instant expiresAt) {}

    public record EnrollmentFinishRequest(
            String ikPub,        // Base64 (X.509 SPKI)
            String kxPub,        // Base64 (X.509 SPKI)
            String bindingSig,   // Base64 (Ed25519 Sign(kxPub-bound))
            String proof,        // Base64 (Ed25519 Sign(enroll-msg))
            String platform,     // optional: 'ios','android','web','desktop'
            String deviceName    // optional
    ) {}

    public record RegisterUserRequest(String handle, String displayName, String password) {}

    public record CreateConversationRequest(boolean isGroup, Set<String> memberHandles) {}
    public record CreateConversationResponse(UUID conversationId, Instant createdAt) {}

    public record AddMemberDevicesRequest(String handle, Set<UUID> deviceIds) {} // Geräte müssen dem User gehören

    public record DistributeCKRequest(
            int epoch,
            Map<UUID,String> sealedForDevice, // deviceId -> base64(ciphertext)
            String sigFromDevice,             // Ed25519 signierte Control-Header (Base64), optional serverseitig ungeprüft
            UUID fromDeviceId
    ) {}

    public record SendMessageRequest(
            String contentType,    // z.B. "text/plain"
            int epoch,
            long counter,
            String ciphertextB64   // AEAD-Ciphertext (gleich für alle Zielgeräte)
    ) {}

    public record SendMessageResponse(UUID messageId, Instant createdAt, int deliveries) {}
}

File: frontend/src/api/dtos.ts
export type UUID = string;
export type ISODateString = string;

export interface DeviceSummary {
  deviceId: UUID;
  deviceName: string;
  platform: string;
  createdAt: ISODateString;
  lastSeenAt: ISODateString;
  revoked: boolean;
}

export interface EnrollmentStartRequest {
  userHandle: string;
}

export interface EnrollmentStartResponse {
  deviceId: UUID;
  nonce: string;
  expiresAt: ISODateString;
}

export interface EnrollmentFinishRequest {
  ikPub: string;
  kxPub: string;
  bindingSig: string;
  proof: string;
  platform?: string;
  deviceName?: string;
}

export interface RegisterUserRequest {
  handle: string;
  displayName: string;
  password: string;
}

export interface CreateConversationRequest {
  isGroup: boolean;
  memberHandles: string[];
}

export interface CreateConversationResponse {
  conversationId: UUID;
  createdAt: ISODateString;
}

export interface AddMemberDevicesRequest {
  handle: string;
  deviceIds: UUID[];
}

export interface DistributeCKRequest {
  epoch: number;
  sealedForDevice: Record<UUID, string>;
  sigFromDevice?: string;
  fromDeviceId: UUID;
}

export interface SendMessageRequest {
  contentType: string;
  epoch: number;
  counter: number;
  ciphertextB64: string;
}

export interface SendMessageResponse {
  messageId: UUID;
  createdAt: ISODateString;
  deliveries: number;
}

export interface BootstrapResponse {
  userId: UUID;
  handle: string;
  conversations: ConversationBrief[];
  devices: UserDeviceBrief[];
}

export interface ConversationBrief {
  conversationId: UUID;
  isGroup: boolean;
  createdAt: ISODateString;
  members: MemberBrief[];
}

export interface MemberBrief {
  userId: UUID;
  handle: string;
}

export interface UserDeviceBrief {
  deviceId: UUID;
  platform: string;
  revokedAt: ISODateString | null;
  lastSeenAt: ISODateString | null;
}

export interface InboxResponse {
  items: DeliveryDTO[];
  nextCursor: string | null;
}

export interface DeliveryDTO {
  deliveryId: UUID;
  messageId: UUID;
  conversationId: UUID;
  contentType: string;
  msgHeaderJson: string;
  ciphertextB64: string;
  createdAt: ISODateString;
}

export interface AckRequest {
  deviceId: UUID;
  deliveryIds: UUID[];
}

export interface ReadRequest {
  deviceId: UUID;
  messageId: UUID;
}

export interface ReadWsMessage {
  conversationId: UUID;
  messageId: UUID;
}

export interface SendWsMessage {
  conversationId: UUID;
  contentType: string;
  epoch: number;
  counter: number;
  ciphertextB64: string;
}

export interface SendAckEvent {
  messageId: UUID;
  conversationId: UUID;
  createdAt: ISODateString;
  deliveries: number;
}

export interface ReadEvent {
  conversationId: UUID;
  messageId: UUID;
  byDeviceId: UUID;
  at: ISODateString;
}

export interface DeliveredEvent {
  messageId: UUID;
  conversationId: UUID;
  recipientDeviceId: UUID;
  createdAt: ISODateString;
}

export interface JwtClaims {
  userId: UUID;
  deviceId: UUID;
}

export interface InboxQueryParams {
  since?: string | null;
  limit?: number;
}

export interface AppUser {
  id: UUID;
  handle: string;
  displayName: string;
  passwordHash: string;
  twoFaEnabled: boolean;
  createdAt: ISODateString;
}

export interface UserResponse {
  id: UUID;
  handle: string;
  displayName: string;
  twoFaEnabled: boolean;
  createdAt: ISODateString;
}
